src/core/reaction_ensemble.cpp

```diff
@@ -0,0 +1,2061 @@
+//method according to smith94x for the reaction ensemble at constant volume and temperature, for the reaction ensemble at constant pressure additionally employ a barostat!
+//NOTE: a chemical reaction consists of a forward and backward reaction. Here both reactions have to be defined seperately.
+//The extent of the reaction is here chosen to be +1.
+//If the reaction trial move for a dissociation of HA is accepted then there is one more dissociated ion pair H+ and A-
+//NOTE: generic_oneway_reaction does not break bonds for simple reactions. as long as there are no reactions like 2A -->B where one of the reacting A particles occurs in the polymer (think of bond breakages if the monomer in the polymer gets deleted in the reaction). This constraint is not of fundamental reason, but there would be a need for a rule for such "collision" reactions (a reaction like the one above).
+//NOTE: paricle types have to start at one and have to increase by one for every type otherwise the function hide_particle cannot work correctly. Through adding 100 in the function hide_particle we ensure that the function works correctly if the particle types are monotonically increasing and if the largest particle type is smaller than 100.
+
+#include "reaction_ensemble.hpp"
+#include "random.hpp" //for random numbers
+#include "energy.hpp"	//for energies
+#include "external_potential.hpp" //for energies
+#include "global.hpp" //for access to global variables
+#include "particle_data.hpp" //for particle creation, modification
+#include "statistics.hpp" //for distto
+#include "integrate.hpp" //for time_step
+#include <stdlib.h>  // for qsort()
+#include <stdio.h> //for getline()
+#include <iostream> //for std::cout
+#include <fstream> //for std::ifstream, std::ofstream for input output into files
+#include "utils.hpp" // for PI
+
```

Comment:
Please put everything in a namespace. You have symbols like replace() that are globally visible and bound to cause collisions.